src/crypto-api/key-passphrase.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2024 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { randomString } from "../randomstring.ts";
+
+const DEFAULT_ITERATIONS = 500000;
+
+const DEFAULT_BITSIZE = 256;
+
+/* eslint-disable camelcase */
+interface IAuthData {
+    private_key_salt?: string;
+    private_key_iterations?: number;
+    private_key_bits?: number;
+}
+/* eslint-enable camelcase */
+
+interface IKey {
+    key: Uint8Array;
+    salt: string;
+    iterations: number;
+}
+
+/**
+ * Derive a key from a passphrase using the salt and iterations from the auth data.
+ * @param authData
+ * @param password
+ */
+export function keyFromAuthData(authData: IAuthData, password: string): Promise<Uint8Array> {
+    if (!authData.private_key_salt || !authData.private_key_iterations) {
+        throw new Error("Salt and/or iterations not found: " + "this backup cannot be restored with a passphrase");
+    }
+
+    return deriveKey(
+        password,
+        authData.private_key_salt,
+        authData.private_key_iterations,
+        authData.private_key_bits || DEFAULT_BITSIZE,
+    );
+}
```

Comment:
Really, this thing ought to be removed; it's about deriving a *backup* key from a *backup* passphrase. That's a thing that we used to do before 4S was a thing, but afaict it was never actually specced, and it's certainly not a normal thing to do to nowadays. There is some code in react-sdk that calls it (via MatrixClient.restoreKeyBackupWithPassword and MatrixClient.keyBackupKeyFromPassword), but I don't think it actually works. https://github.com/element-hq/element-web/issues/27455#issuecomment-2160771635).

Still, we're not really ready to rip it out just yet. Since it's not currently used directly from outside the js-sdk, I suggest that rather than exposing a new method to applications via `crypto-api`, you instead put this somewhere within `common-crypto`.